\documentclass[table,xcolor=dvipsnames]{beamer}


\usepackage{xcolor}
\usepackage{url}
\usepackage{amsmath}
\usepackage{amsthm}
\usepackage{amssymb}
\usepackage{graphicx}
\usepackage{multicol}
\usepackage{tikz}
\usetikzlibrary{shapes,arrows}
\usepackage{graphicx}
\usepackage{float}
\usepackage{hyperref}
\usepackage{multicol}
\usepackage{multirow}
\usepackage{/u/w/e/welch/R-3.1.0/share/texmf/tex/latex/Sweave}

\usepackage{pifont}% http://ctan.org/pkg/pifont




\usetheme{Szeged}
\usecolortheme[named=RoyalBlue]{structure}
%\usecolortheme[accent=blue]{solarized}


% Personal Data
\title{How to use the profile package}
\date{April 2014}
\author{Rene Welch}
\institute{University of Wisconsin - Madison}


\setbeamertemplate{sections/subsections in toc}[square] % This is to remove the symbols of the slides
\setbeamertemplate{navigation symbols}{} % This is to convert the "3d" spheres to squares on toc
\setbeamertemplate{itemize item}[square] % This is to convert the "3d" spheres to squares on enumerate



\begin{document}

<<<setup, include=FALSE, cache=FALSE>>=
library(knitr)
knit_hooks$set(packages = function(before, options, envir) {
  if (before) {
    ## load packages before a chunk is executed
    for (p in options$packages) library(p, character.only = TRUE)
  }
})
figsdir = "figs/generated"
@ 


\begin{frame}
  \maketitle
\end{frame}

\begin{frame}
\tableofcontents
\end{frame}


\section{A typical ChIP-Seq experiment}
\begin{frame}
\frametitle{A typical ChIP-Seq experiment}

\begin{figure}
  \includegraphics[width = .75\textwidth]{figs/fromPapers/parkOverview.png}
\caption{From Park's ChIP-seq: Advantages and challenges of a maturing technology} 
\end{figure}
\end{frame}


\begin{frame}


\only<1>{
\begin{itemize}
\item The double stranded {\color{RoyalBlue} DNA}
\end{itemize}}
\only<2>{
\begin{itemize}
\item There is an interaction with a protein (TF or histone)
\end{itemize}}
\only<3>{
\begin{itemize}
\item The protein is removed by an antibody
\end{itemize}}
\only<4-6>{
\begin{itemize}
\item The DNA is fragmented by sonication
\end{itemize}}

\begin{figure}[h!]
\centering
\begin{tikzpicture}

\uncover<1->{\draw [line width=10,color=red] (0,1) -- (10,1); 
             \node [color=red] at (-.5,1) {5'};
             \node [color=red] at (10.5,1) {3'};
             \draw [line width=10,color=blue] (0,0) -- (10,0);
             \node [color=blue] at (-.5,0) {3'};
             \node [color=blue] at (10.5,0) {5'};}
\uncover<2>{\draw [color=yellow,fill=yellow] (5,1.5) circle [radius=1.5];
             \node at (5,2) {\huge{Protein}};}
\uncover<3-4>{\draw [color=yellow,fill=yellow] (5,3.5) circle [radius=1.5];
             \node at (5,4) {\huge{Protein}};}
\uncover<4->{\draw [->,line width=3,color=black] (2,-1) -- (2,0);
             \draw [->,line width=3,color=black] (3,-1) -- (3,0);
             \draw [->,line width=3,color=black] (7.5,-1) -- (7.5,0);
             \draw [->,line width=3,color=black] (6.7,-1) -- (6.7,0);
             \draw [line width=2,color=black,dashed] (2,-1) -- (2,1);
             \draw [line width=2,color=black,dashed] (3,-1) -- (3,1);
             \draw [line width=2,color=black,dashed] (7.5,-1) -- (7.5,1);
             \draw [line width=2,color=black,dashed] (6.7,-1) -- (6.7,1);
             \node [color=black] at (9.5,-.5) {\large{Fragmentation}};}
\uncover<5->{\draw [line width=3,color=red] (2,1.5) -- (4,1.5);
             \draw [line width=3,color=red] (3,1.8) -- (5,1.8);
             \draw [line width=3,color=blue] (5.5,1.5) -- (7.5,1.5);
             \draw [line width=3,color=blue] (4.7,2.3) -- (6.7,2.3);}
\uncover<6->{\draw [line width=3,color=red] (3.3,2) -- (5.3,2);
             \draw [line width=3,color=red] (2.5,1.6) -- (4.5,1.6);
             \draw [line width=3,color=red] (3.7,2.2) -- (5.7,2.2);
             \draw [line width=3,color=blue] (5.3,1.7) -- (7.3,1.7);
             \draw [line width=3,color=blue] (4.5,2.5) -- (6.5,2.5);}

\end{tikzpicture}
\end{figure}
\end{frame}



<<sim, packages=c('GenomicRanges', 'ggplot2'),include=FALSE,echo=FALSE,eval=FALSE>>=
peaksim <- function(nreads,mu,p,delta,sigma2,readLength)
{  
  strand = runif(nreads) >  p  # TRUE means Fwd, FALSE means Bwd
  fwd_pos = sum(strand)
  bwd_pos = nreads - fwd_pos
  fwd_start = round(rnorm(fwd_pos,mu - delta, sd = sqrt(sigma2)),0)
  bwd_start = round(rnorm(bwd_pos,mu + delta, sd = sqrt(sigma2)),0)
  peakReads = c(
    GRanges(seqnames = "sim",ranges = IRanges(start = fwd_start,width = readLength),strand = "+"),
    GRanges(seqnames = "sim",ranges = IRanges(end = bwd_start,width = readLength),strand = "-")
    )  
  return(peakReads)
}

plotPeak <- function(peak,m)
{
  fwd_reads = suppressWarnings(subset(peak,subset = strand(peak)=="+"))
  bwd_reads = suppressWarnings(subset(peak,subset = strand(peak)=="-"))
  fwd_cov = coverage(fwd_reads)[[1]]
  bwd_cov = coverage(bwd_reads)[[1]]  
  x = 1:m
  fwd_xp = runLength(fwd_cov)[1:(nrun(fwd_cov))]
  fwd_yp = c(runValue(fwd_cov) ,0)
  fwd_y = stepfun(cumsum(fwd_xp),fwd_yp)(x)
  bwd_xp = runLength(bwd_cov)[1:(nrun(bwd_cov))]
  bwd_yp = c(runValue(bwd_cov),0)
  bwd_y = stepfun(cumsum(bwd_xp),bwd_yp)(x)
  fwd = data.frame(coord = x,counts = fwd_y,strand = "fwd")
  bwd = data.frame(coord = x,counts = bwd_y,strand = "bwd")
  df = rbind(fwd,bwd)
#  mm = max(c(fwd_y,bwd_y))
#  yl = 1.2 * c(0,mm)
  df$strand = factor(df$strand,levels = c("fwd","bwd"))
  peakPlot = ggplot(df,aes(coord,counts,colour = strand))+geom_line()+scale_colour_manual(values = c("red","blue"))+
    theme(legend.position = "top",text = element_text(size = 20))+xlab("genomic coordinate")
  return(peakPlot)
}  

nreads = 1000
mu = 250
p = .5
delta = 75
sigma2 = 2350
readLength = 51

peakReads = peaksim(nreads,mu,p,delta,sigma2,readLength)
pdf(file = file.path(figsdir,"peak.pdf"),width =12,height = 6)
p = plotPeak(peakReads,500)
print(p)
dev.off()
@ 


\begin{frame}
\frametitle{A typical (simulated) peak}

\begin{figure}
\includegraphics[width = .8\textwidth]{figs/generated/peak.pdf}
\end{figure}

\end{frame}




\section{How to use the profile package?}
\subsection{How to input data}
\begin{frame}[t]
\frametitle{How to input data}

For this example we are going to load the package and parameters:

<<howtoload,include=FALSE,echo=FALSE,eval=TRUE>>=
  library(profile)
  source("../rscripts/all_functions.R")  
@ 

<<outputload,include=TRUE,tidy=TRUE>>=
  library(profile)

  regionName = "CTCF-K562"

  fileFormat = "bam"

  maxBandwidth = 301

  bandwidth = 151

  fragLen = 200

  remChr = "chrY"

  windowExt = 2000

  mc.cores = 7
@ 

\end{frame}


\end{document}
